Add permit-like function to support gasless withdrawals

gibraltariq opened the pull request:
I may be oversimplifying this but it seems the only thing we need to support gasless withdrawals would be permission to move the ERC 20 tokens for the owner. The contract can already burn the ERC 1155 token for the owner upon withdraw without a permit.

I am building on top of [Pranesh's change](https://github.com/enclosed-money/contracts/pull/13) for the testing so I'll properly merge this when it's approved and merged.

gibraltariq committed:
Copy of overlap - Add depositWithPermit() method

gibraltariq committed:
Add permit function.

gibraltariq committed:
Add gas snapshot.

w1nt3r-eth commented on src/NFTBill.sol:
```diff
@@ -39,6 +40,29 @@ contract NFTBill is ERC1155 {
         _mint(msg.sender, id, 1, '');
     }
 
+    // To do a gasless withdraw we just need permission to transfer the NFT owner's
+    // ERC20 (and we could take some of it to pay the gas fees).
```
Well this can allow gasless deposit, not withdraw! This allows the user to skip calling `approve` on ERC20 before calling deposit.

gibraltariq replied:
Updated! Hopefully got it right/closer this time

gibraltariq committed:
Fix - add permit-like approval for gasless withdrawal.

gibraltariq committed:
Fix - update snapshots.

gibraltariq commented:
Will add a negative test case and fuzzing once the main logic is approved.

gibraltariq committed:
Use _operatorApprovals included with standard ERC1155 instead of our own allowances.

gibraltariq committed:
Update snapshot.

gibraltariq committed:
Add withdrawFrom to support gasless withdraws.

gibraltariq commented:
Added a couple changes:

1. Switched to using operatorApprovals instead of allowances per your advice @w1nt3r-eth 
2. Implemented a check for approval in withdrawFrom. This can be merged with withdrawTo in [this PR](https://github.com/enclosed-money/contracts/pull/15) depending on whichever one gets merged in first

stoneworld commented:
just add function withdrawWithPermit is more concise.
